<a href="https://colab.research.google.com/github/janhavi-lab/GreenTechi/blob/main/Electricity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


In [2]:
df = pd.read_csv('/content/classification_dataset.csv')

In [3]:
X = df.drop('Wastage_Label', axis=1)
y = df['Wastage_Label']
display(X.head())
display(y.head())

RoomNo  Month  Day  Hour  Consumption_kWh  AverageConsumption_kWh  \
0       1      8    1     0            0.882                   0.342   
1       1      8    1     1            0.137                   0.423   
2       1      8    1     2            0.106                   0.244   
3       1      8    1     3            0.110                   0.373   
4       1      8    1     4            0.090                   0.585   

   Wastage_kWh  
0         0.54  
1         0.00  
2         0.00  
3         0.00  
4         0.00

0    Medium Wastage Room
1             Green Room
2             Green Room
3             Green Room
4             Green Room
Name: Wastage_Label, dtype: object

In [4]:
display(df.head())
display(df.info())
display(df.describe())

RoomNo  Month  Day  Hour  Consumption_kWh  AverageConsumption_kWh  \
0       1      8    1     0            0.882                   0.342   
1       1      8    1     1            0.137                   0.423   
2       1      8    1     2            0.106                   0.244   
3       1      8    1     3            0.110                   0.373   
4       1      8    1     4            0.090                   0.585   

   Wastage_kWh        Wastage_Label  
0         0.54  Medium Wastage Room  
1         0.00           Green Room  
2         0.00           Green Room  
3         0.00           Green Room  
4         0.00           Green Room

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2520 entries, 0 to 2519
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   RoomNo                  2520 non-null   int64  
 1   Month                   2520 non-null   int64  
 2   Day                     2520 non-null   int64  
 3   Hour                    2520 non-null   int64  
 4   Consumption_kWh         2520 non-null   float64
 5   AverageConsumption_kWh  2520 non-null   float64
 6   Wastage_kWh             2520 non-null   float64
 7   Wastage_Label           2520 non-null   object 
dtypes: float64(3), int64(4), object(1)
memory usage: 157.6+ KB


None

RoomNo   Month          Day        Hour  Consumption_kWh  \
count  2520.000000  2520.0  2520.000000  2520.00000      2520.000000   
mean      8.000000     8.0     4.000000    11.50000         0.484710   
std       4.321351     0.0     2.000397     6.92356         0.478364   
min       1.000000     8.0     1.000000     0.00000         0.051000   
25%       4.000000     8.0     2.000000     5.75000         0.133000   
50%       8.000000     8.0     4.000000    11.50000         0.218000   
75%      12.000000     8.0     6.000000    17.25000         0.824750   
max      15.000000     8.0     7.000000    23.00000         1.713000   

       AverageConsumption_kWh  Wastage_kWh  
count             2520.000000  2520.000000  
mean                 0.404694     0.234438  
std                  0.114243     0.377046  
min                  0.200000     0.000000  
25%                  0.307000     0.000000  
50%                  0.408000     0.000000  
75%                  0.502000     0.416000  
max                  0.600000     1.481000

In [5]:
df.isnull().sum()

RoomNo                    0
Month                     0
Day                       0
Hour                      0
Consumption_kWh           0
AverageConsumption_kWh    0
Wastage_kWh               0
Wastage_Label             0
dtype: int64

In [6]:
df.drop_duplicates()

RoomNo  Month  Day  Hour  Consumption_kWh  AverageConsumption_kWh  \
0          1      8    1     0            0.882                   0.342   
1          1      8    1     1            0.137                   0.423   
2          1      8    1     2            0.106                   0.244   
3          1      8    1     3            0.110                   0.373   
4          1      8    1     4            0.090                   0.585   
...      ...    ...  ...   ...              ...                     ...   
2515      15      8    7    19            0.051                   0.572   
2516      15      8    7    20            1.430                   0.497   
2517      15      8    7    21            1.562                   0.203   
2518      15      8    7    22            0.058                   0.412   
2519      15      8    7    23            0.450                   0.542   

      Wastage_kWh        Wastage_Label  
0           0.540  Medium Wastage Room  
1           0.000           Green Room  
2           0.000           Green Room  
3           0.000           Green Room  
4           0.000           Green Room  
...           ...                  ...  
2515        0.000           Green Room  
2516        0.933    High Wastage Room  
2517        1.359    High Wastage Room  
2518        0.000           Green Room  
2519        0.000           Green Room  

[2520 rows x 8 columns]

In [7]:
X

RoomNo  Month  Day  Hour  Consumption_kWh  AverageConsumption_kWh  \
0          1      8    1     0            0.882                   0.342   
1          1      8    1     1            0.137                   0.423   
2          1      8    1     2            0.106                   0.244   
3          1      8    1     3            0.110                   0.373   
4          1      8    1     4            0.090                   0.585   
...      ...    ...  ...   ...              ...                     ...   
2515      15      8    7    19            0.051                   0.572   
2516      15      8    7    20            1.430                   0.497   
2517      15      8    7    21            1.562                   0.203   
2518      15      8    7    22            0.058                   0.412   
2519      15      8    7    23            0.450                   0.542   

      Wastage_kWh  
0           0.540  
1           0.000  
2           0.000  
3           0.000  
4           0.000  
...           ...  
2515        0.000  
2516        0.933  
2517        1.359  
2518        0.000  
2519        0.000  

[2520 rows x 7 columns]

In [8]:
y

0       Medium Wastage Room
1                Green Room
2                Green Room
3                Green Room
4                Green Room
               ...         
2515             Green Room
2516      High Wastage Room
2517      High Wastage Room
2518             Green Room
2519             Green Room
Name: Wastage_Label, Length: 2520, dtype: object

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (2016, 7)
X_test shape: (504, 7)
y_train shape: (2016,)
y_test shape: (504,)


In [11]:
from sklearn.preprocessing import LabelEncoder

# Encode the target variable
label_encoder = LabelEncoder()
y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

model = LinearRegression()
model.fit(X_train, y_train_encoded)

LinearRegression()

In [14]:
y_pred = model.predict(X_test)

# Evaluate the model
print('Mean squared error (MSE):', mean_squared_error(y_test_encoded, y_pred))
print('Coefficient of determination (R^2):', r2_score(y_test_encoded, y_pred))

Mean squared error (MSE): 0.252679502429122
Coefficient of determination (R^2): 0.5635093780285496


# Task
Compare the performance of Logistic Regression, Decision Tree, and Random Forest classification models on the provided dataset and identify the best-performing model.

## Import necessary libraries

### Subtask:
Import the required classification models and evaluation metrics from scikit-learn.


**Reasoning**:
Import the required classification models and evaluation metrics for the subtask.



In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

## Train and evaluate logistic regression

### Subtask:
Train a Logistic Regression model on the training data and evaluate its performance on the test data using metrics like accuracy, precision, recall, and F1-score.


**Reasoning**:
Train a Logistic Regression model and evaluate its performance.



In [16]:
logistic_regression_model = LogisticRegression(random_state=42, solver='liblinear')
logistic_regression_model.fit(X_train, y_train_encoded)
y_pred_lr = logistic_regression_model.predict(X_test)

accuracy_lr = accuracy_score(y_test_encoded, y_pred_lr)
precision_lr = precision_score(y_test_encoded, y_pred_lr, average='weighted')
recall_lr = recall_score(y_test_encoded, y_pred_lr, average='weighted')
f1_lr = f1_score(y_test_encoded, y_pred_lr, average='weighted')

print(f"Logistic Regression Accuracy: {accuracy_lr}")
print(f"Logistic Regression Precision: {precision_lr}")
print(f"Logistic Regression Recall: {recall_lr}")
print(f"Logistic Regression F1-score: {f1_lr}")

Logistic Regression Accuracy: 0.9444444444444444
Logistic Regression Precision: 0.9485448025363992
Logistic Regression Recall: 0.9444444444444444
Logistic Regression F1-score: 0.9398811295713088


## Train and evaluate decision tree

### Subtask:
Train a Decision Tree model on the training data and evaluate its performance on the test data using the same classification metrics.


**Reasoning**:
Train a Decision Tree model and evaluate its performance on the test data.



In [17]:
decision_tree_model = DecisionTreeClassifier(random_state=42)
decision_tree_model.fit(X_train, y_train_encoded)
y_pred_dt = decision_tree_model.predict(X_test)

accuracy_dt = accuracy_score(y_test_encoded, y_pred_dt)
precision_dt = precision_score(y_test_encoded, y_pred_dt, average='weighted')
recall_dt = recall_score(y_test_encoded, y_pred_dt, average='weighted')
f1_dt = f1_score(y_test_encoded, y_pred_dt, average='weighted')

print(f"Decision Tree Accuracy: {accuracy_dt}")
print(f"Decision Tree Precision: {precision_dt}")
print(f"Decision Tree Recall: {recall_dt}")
print(f"Decision Tree F1-score: {f1_dt}")

Decision Tree Accuracy: 1.0
Decision Tree Precision: 1.0
Decision Tree Recall: 1.0
Decision Tree F1-score: 1.0


## Train and evaluate random forest

### Subtask:
Train a Random Forest model on the training data and evaluate its performance on the test data using the same classification metrics.


**Reasoning**:
Train a Random Forest model and evaluate its performance on the test data using the specified metrics.



In [18]:
random_forest_model = RandomForestClassifier(random_state=42)
random_forest_model.fit(X_train, y_train_encoded)
y_pred_rf = random_forest_model.predict(X_test)

accuracy_rf = accuracy_score(y_test_encoded, y_pred_rf)
precision_rf = precision_score(y_test_encoded, y_pred_rf, average='weighted')
recall_rf = recall_score(y_test_encoded, y_pred_rf, average='weighted')
f1_rf = f1_score(y_test_encoded, y_pred_rf, average='weighted')

print(f"Random Forest Accuracy: {accuracy_rf}")
print(f"Random Forest Precision: {precision_rf}")
print(f"Random Forest Recall: {recall_rf}")
print(f"Random Forest F1-score: {f1_rf}")

Random Forest Accuracy: 0.998015873015873
Random Forest Precision: 0.9980419799498746
Random Forest Recall: 0.998015873015873
Random Forest F1-score: 0.9980164304667616


## Compare model performance

### Subtask:
Summarize and compare the evaluation metrics of the three models to determine which one performed best.


**Reasoning**:
Print the evaluation metrics for each model and then create a summary table to compare their performance.



In [20]:
print("--- Model Performance Comparison ---")
print("\nLogistic Regression:")
print(f"Accuracy: {accuracy_lr:.4f}")
print(f"Precision: {precision_lr:.4f}")
print(f"Recall: {recall_lr:.4f}")
print(f"F1-score: {f1_lr:.4f}")

print("\nDecision Tree:")
print(f"Accuracy: {accuracy_dt:.4f}")
print(f"Precision: {precision_dt:.4f}")
print(f"Recall: {recall_dt:.4f}")
print(f"F1-score: {f1_dt:.4f}")

print("\nRandom Forest:")
print(f"Accuracy: {accuracy_rf:.4f}")
print(f"Precision: {precision_rf:.4f}")
print(f"Recall: {recall_rf:.4f}")
print(f"F1-score: {f1_rf:.4f}")

# Create a summary table
performance_summary = pd.DataFrame({
    'Metric': ['Accuracy', 'Precision', 'Recall', 'F1-score'],
    'Logistic Regression': [accuracy_lr, precision_lr, recall_lr, f1_lr],
    'Decision Tree': [accuracy_dt, precision_dt, recall_dt, f1_dt],
    'Random Forest': [accuracy_rf, precision_rf, recall_rf, f1_rf]
})

print("\n--- Performance Summary Table ---")
display(performance_summary)

best_model = performance_summary.set_index('Metric').idxmax(axis=1)
print("\n--- Best Performing Model per Metric ---")
display(best_model)

# Overall best model based on average F1-score (or another chosen metric)
overall_best_model = performance_summary.set_index('Metric').loc['F1-score'].idxmax()
print(f"\nOverall best model based on F1-score: {overall_best_model}")

--- Model Performance Comparison ---

Logistic Regression:
Accuracy: 0.9444
Precision: 0.9485
Recall: 0.9444
F1-score: 0.9399

Decision Tree:
Accuracy: 1.0000
Precision: 1.0000
Recall: 1.0000
F1-score: 1.0000

Random Forest:
Accuracy: 0.9980
Precision: 0.9980
Recall: 0.9980
F1-score: 0.9980

--- Performance Summary Table ---


Metric  Logistic Regression  Decision Tree  Random Forest
0   Accuracy             0.944444            1.0       0.998016
1  Precision             0.948545            1.0       0.998042
2     Recall             0.944444            1.0       0.998016
3   F1-score             0.939881            1.0       0.998016


--- Best Performing Model per Metric ---


Metric
Accuracy     Decision Tree
Precision    Decision Tree
Recall       Decision Tree
F1-score     Decision Tree
dtype: object


Overall best model based on F1-score: Decision Tree


## Summary:

### Data Analysis Key Findings

*   The Logistic Regression model achieved an accuracy of approximately 0.9444, precision of 0.9485, recall of 0.9444, and an F1-score of 0.9399 on the test data.
*   The Decision Tree model demonstrated perfect performance on the test set with an accuracy, precision, recall, and F1-score of 1.0.
*   The Random Forest model also performed exceptionally well, with accuracy, precision, recall, and F1-score all approximately 0.998.
*   Based on the evaluation metrics (Accuracy, Precision, Recall, and F1-score), the Decision Tree model was the best-performing model in this comparison, achieving a perfect score across all metrics.

### Insights or Next Steps

*   The perfect scores achieved by the Decision Tree and near-perfect scores by the Random Forest on the test set might indicate potential overfitting or a very separable dataset. Further investigation using cross-validation or a larger, independent test set could provide a more robust evaluation.
*   Given the high performance of both tree-based models, exploring hyperparameter tuning for the Decision Tree and Random Forest models could potentially improve their generalization ability and confirm their performance on unseen data.
